app/src/main/java/tech/ula/utils/ValidationUtility.kt

```diff
@@ -1,31 +1,86 @@
+    fun validateUsername(username: String): Credential {
+        var usernameIsValid = false
+        var errorMessage = ""
+        val blacklist = resources.getStringArray(R.array.blacklisted_usernames)
 
-        pattern = Pattern.compile(USERNAME_REGEX)
-        matcher = pattern.matcher(password)
+        when {
+            username.isEmpty() -> {
+                errorMessage = resources.getString(R.string.error_empty_field)
+            }
+            !validateUsernameCharacters(username) -> {
+                errorMessage = resources.getString(R.string.error_username_invalid_characters)
+            }
+            blacklist.contains(username) -> {
+                errorMessage = resources.getString(R.string.error_username_in_blacklist)
+                errorMessage = String.format(errorMessage, username)
+            }
+            else ->
+                usernameIsValid = true
+        }
 
-        return matcher.matches()
+        return Credential(usernameIsValid, errorMessage)
     }
 
-    fun isPasswordValid(password: String): Boolean {
+    fun validatePasswords(password: String, vncPassword: String): Credential {
+        var passwordIsValid = false
+        var errorMessage = ""
+
+        when {
+            password.isEmpty() || vncPassword.isEmpty() -> {
+                errorMessage = resources.getString(R.string.error_empty_field)
+            }
+            vncPassword.length > 8 || vncPassword.length < 6 -> {
+                errorMessage = resources.getString(R.string.error_vnc_password_length_incorrect)
+            }
+            !validatePasswordCharacters(password) -> {
+                errorMessage = resources.getString(R.string.error_password_invalid)
+            }
+            !validatePasswordCharacters(vncPassword) -> {
+                errorMessage = resources.getString(R.string.error_vnc_password_invalid)
+            }
+            else ->
+                passwordIsValid = true
+        }
+
+        return Credential(passwordIsValid, errorMessage)
+    }
+
+    private fun validateUsernameCharacters(username: String): Boolean {
+        val usernameRegex = "([a-z_][a-z0-9_]{0,30})"
+
+        val compiledRegex = Pattern.compile(usernameRegex)
+        val matcher = compiledRegex.matcher(username)
+        val hasValidCharacters = matcher.matches()
+
+        if (hasValidCharacters) {
+            return true
+        }
+        return false
+    }
+
+    private fun validatePasswordCharacters(password: String): Boolean {
         val pattern: Pattern
         val matcher: Matcher
 
-        val PASSWORD_REGEX = "^[a-zA-Z0-9!@#$%^&*()_+=,./?<>:]*\$"
+        val passwordRegex = "^[a-zA-Z0-9!@#$%^&*()_+=,./?<>:]*\$"
 
-        pattern = Pattern.compile(PASSWORD_REGEX)
+        pattern = Pattern.compile(passwordRegex)
         matcher = pattern.matcher(password)
 
         return matcher.matches()
     }
-}
\ No newline at end of file
+}
+
+sealed class CredentialValidation
```

Comment:
There's no real need for this sealed class here. Data classes can exist without inheriting from a sealed class. I also think the name could be more clear, maybe `CredentialValidationStatus`, or something?

Reply:
I didn't know that data classes could stand on its own.  I've removed the unnecessary sealed class